vm/tests/grammar.rs

```diff
@@ -715,6 +715,24 @@ fn peek_all() {
     };
 }
 
+#[test]
+fn peek_slice_23() {
+    parses_to! {
+        parser: vm(),
+        input: "0123421",
```

Comment:
How about: `0..-1` matches from bottom to top and `-1..0` matches from top to bottom? You should at all times be able to tell whether the indices are inverted, in which case you would use it the way it is implemented right now.

Reply:
OK, so in `[a..b]`, when

- a < b (after normalizing), we match `stack_vec[a..b]` (bottom to top / push order)
- a > b (after normalizing), we match `stack_vec[b..a].rev()` (top to bottom / pop order)

We introduce `-0` as the top (`stack_vec.len()`), otherwise there’s no way to say “match from top to X”. With `-0`, that’s `[-0..1]`

We should also make both sides mandatory then. If the numbers indicate direction, empty sides being specific to one direction will be confusing (`[-0..]` would mean top-to-top)

Reply:
-1 is the index before the last item. Basically:

```
[ A, B, C]
  0  1  2  3
 -3 -2 -1 -0
```

And

Reply:
At least how it's done in Python:

```
  0  1  2  3  4
[  A  B  C  D  ]
 -4 -3 -2 -1 -0
```

As such `list[0:3]` takes the list starting at `0` and ending at `3`, `[A B C]`. The index for slicing refers to the space between the elements.

I think semantics of the slice very much should _not_ depend on the length of the stack. If I'm slicing from `[-2..2]`, I don't know what I'd really expect, but it's definitely not "the whole thing forwards if it's two elements long, the middle minus the ends reversed if it's four or more".

It's for this reason that backwards slicing is a hard problem to solve, _especially_ when mixing unknown and/or negative indexing into the question.

Reply:
Per the above diagram, `list[-1]` is the last element. If you want to slice off the last element, you'd do `list[s..-1]`, and if you want to slice to the end, you'd do `list[s..]`.

I think that the most correct behavior would be for `pest::stack::peek[..]` to match the entire stack as is, in the order they were pushed, and if you want to slice to the end, leave off the final index instead of e.g. `[..-0]`.

I also think allowing a single index instead of a range is unnecessary and serves to complicate the issue. Part of the complexity is that slicing operates with the indexes referring to the space inbetween elements, wheras indexing necessarily has the index refer to the element itself.

Reply:
pest slicing could be defined as the inclusive range between two indices, always slicing everything forward, order of the indices be damned. This behavior would be different from Rust's ranges, though.

Reply:
This also fits with the current definition of character ranges which are inclusive.

Reply:
I'd generally lean for a more obscure/different approach, as long as it's the most coherent and with the least assumptions. And this is because this features will be somewhat obscure itself, so learning to use it properly will be something the user is expected to do anyway.

Reply:
OK, @dragostis, your confusion shows we need a simpler thing, and @CAD97, I agree with you optinions, specifically

> I think semantics of the slice very much should not depend on the length of the stack

I think it’s a bit unfortunate that `PEEK_ALL` matches in the opposite direction as `PEEK` will, but I agree that “push order/bottom to top” is the most intuitive one. Vecs in rust can be used as stacks, and are iterated bottom to top when seen as such.

So I’ll simply leave out everything complicated right now and just invert the matching order for `PEEK[..]`

---

To clear up the -0 thing @dragostis: An index doesn’t point to an element, but between them. `arr[i]` in rust/c/java/… denotes the *start* of an element: `arr[1]` means “The memory location after the first element and before the second”, just like “1cm” on a ruler is the location after the first centimeter and before the second. “-1” is therefore the location before the last element, and “-0” the location after.

In a `..` range, you use those locations like this: `arr[1..2]` means “the memory that covers the second element”, `arr[-1..-0]` means “the memory that covers the last element”.

Reply:
@flying-sheep, of course. Let me develop my proposal a bit. I'll try to make it less hasty.

We can use indices to refer to actual elements, rather than the space between them. Indexing can start from the bottom of the stack with positive (but not strictly) integers 0, 1, 2, ..., or from the top of the stack with strictly negative integers -1, -2, -3, ... in order to avoid having to deal with non-integer primitives. This can be used to index into an element.

For slicing, we can simply use inclusive ranges, i.e. a range will include both the start and end indices and everything between them.

This is less intuitive, but seems to have less of the pitfalls. Anyway, slicing primitives are definitely way harder than they look.

About `PEEK_ALL`: I think the best solution here is to deprecate it right away in `2.1`.

Reply:
Ah I see! sorry for assuming you were confused!

I think the problem with your proposal is that it makes pest look similar enough to Rust slicing syntax to confuse people with its different semantics. Possible solutions (we can do both together):

1. use a different delimiter syntax, e.g. `PEEK[a:b]` | `PEEK[a..=b]` | `PEEK[a-->b]]`.
2. use explicit end syntax, e.g. `PEEK[1..$]` and `PEEK[..$-2]`

The second one maintains the indexes-are-between-elements intuition that most system language programmers probably have.

> Anyway, slicing primitives are definitely way harder than they look.

haha, yeah! designing things to be functional and intuitive is hard!

Reply:
We also still have the problem that mixing negative and positive indices, the order of end and start index is dependent on the length of the stack, which should be avoided. So we have two options here:

1. We don’t decide on a syntax where the matching order depends on the order of indices (such as `PEEK[bottom-->top]` / `PEEK[bottom<--top]` in which the delimiter defines matching order)
2. we don’t allow mixing negative and positive indices. I think “everything except first/last” is too common to do that

Reply:
so what do you say?

Reply:
I would start by adding no syntax for reversed slicing. This can be decided in #333 when we reach a conclusion.

The negative syntax, I think, would be the most intuitive, but I don't have a very strong opinion here.

Reply:
OK, so basically this can be pulled as is?

Reply:
I'll take another look!